man/Planimeter.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "PLANIMETER 1"
.TH PLANIMETER 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
Planimeter \-\- compute the area of geodesic polygons
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBPlanimeter\fR [ \fB\-r\fR ] [ \fB\-s\fR ] [ \fB\-l\fR ] [ \fB\-e\fR \fIa\fR \fIf\fR ]
[ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ] [ \fB\-G\fR | \fB\-Q\fR | \fB\-R\fR ] [ \fB\-E\fR ]
[ \fB\-\-geoconvert\-input\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
Measure the area of a geodesic polygon.  Reads polygon vertices from
standard input, one per line.  Vertices are be given as latitude and
longitude.  By default latitude precedes longitude, however this
convention is reversed with the \fB\-w\fR flag and a hemisphere designator
(\fIN\fR, \fIS\fR, \fIE\fR, \fIW\fR) can be used to disambiguate the coordinates.
The end of input, a blank line, or a line which can\*(Aqt be interpreted
as a vertex signals the end of one polygon and the start of the next.
For each polygon print a summary line with the number of points, the
perimeter (in meters), and the area (in meters^2).
.PP
The edges of the polygon are given by the \fIshortest\fR geodesic (or
rhumb line) between consecutive vertices.  In certain cases, there may
be two or many such shortest path, and in that case, the polygon is
not uniquely specified by its vertices.  For geodesics, this only
happens with very long edges (for the WGS84 ellipsoid, any edge
shorter than 19970 km is uniquely specified by its end points).  In
such cases, insert an additional vertex near the middle of the long
edge to define the boundary of the polygon.
.PP
By default, polygons traversed in a counter\-clockwise direction return a
positive area and those traversed in a clockwise direction return a
negative area.  This sign convention is reversed if the \fB\-r\fR option is
given.
.PP
Of course, encircling an area in the clockwise direction is equivalent
to encircling the rest of the ellipsoid in the counter\-clockwise
direction.  The default interpretation used by \fBPlanimeter\fR is the one
that results in a smaller magnitude of area; i.e., the magnitude of the
area is less than or equal to one half the total area of the ellipsoid.
If the \fB\-s\fR option is given, then the interpretation used is the one
that results in a positive area; i.e., the area is positive and less
than the total area of the ellipsoid.
.PP
Arbitrarily complex polygons are allowed.  In the case of
self\-intersecting polygons the area is accumulated "algebraically",
e.g., the areas of the 2 loops in a figure\-8 polygon will partially
cancel.  Polygons may include one or both poles.  There is no need to
close the polygon.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-r\fR 4
.IX Item "-r"
toggle whether counter\-clockwise traversal of the polygon returns a
positive (the default) or negative result.
.IP \fB\-s\fR 4
.IX Item "-s"
toggle whether to return a signed result (the default) or not.
.IP \fB\-l\fR 4
.IX Item "-l"
toggle whether the vertices represent a polygon (the default) or a
polyline.  For a polyline, the number of points and the length of the
path joining them is returned; the path is not closed and the area is
not reported.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the WGS84 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.  If entering vertices as UTM/UPS or
MGRS coordinates, use the default ellipsoid, since the conversion of
these coordinates to latitude and longitude always uses the WGS84
parameters.
.IP \fB\-w\fR 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
when reading geographic coordinates, longitude precedes latitude (this
can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR, \fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 6); the perimeter is given
(in meters) with \fIprec\fR digits after the decimal point; the area is
given (in meters^2) with (\fIprec\fR \- 5) digits after the decimal point.
.IP \fB\-G\fR 4
.IX Item "-G"
the edges joining the vertices are geodesics.  This is the default
option and is recommended for terrestrial applications.  This option,
\&\fB\-G\fR, and the following two options, \fB\-Q\fR and \fB\-R\fR, are mutually
exclusive.
.IP \fB\-Q\fR 4
.IX Item "-Q"
map the points to the authalic sphere and compute the area of the
resulting spherical polygon.  The area will be reasonable accurate
\&\fIprovided\fR that the edges are sufficiently short.  The perimeter
calculation is not accurate.
.IP \fB\-R\fR 4
.IX Item "-R"
the edges joining the vertices are rhumb lines instead of geodesics.
.IP \fB\-E\fR 4
.IX Item "-E"
use the exact equations for the geodesic \fB\-G\fR, authalic \fB\-Q\fR, and
rhumb \fB\-R\fR calculations instead of series expansions.  For the
geodesic and rhumb methods, the area is computed by applying discrete
sine transforms to the integrand in the expression for the area.
These are more accurate, albeit slower, than the (default) series
expansions for |\fIf\fR| > 0.02 and will give high accuracy for \-99
< \fIf\fR < 0.99.  It is not necessary to specify this option for
terrestrial applications.
.IP \fB\-\-geoconvert\-input\fR 4
.IX Item "--geoconvert-input"
The input lines are interpreted in the same way as \fBGeoConvert\fR\|(1)
allowing the coordinates for the vertices to be given as UTM/UPS or
MGRS coordinates, as well as latitude and longitude.  \fBCAUTION\fR:
GeoConvert assumes the coordinates refer to the WGS84 ellipsoid
(disregarding the \fB\-e\fR flag) and MGRS coordinates signify the center
of the corresponding MGRS square.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing.  For a given polygon, the last such string found will be
appended to the output line (separated by a space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH EXAMPLES
.IX Header "EXAMPLES"
Example (the area of the 100km MGRS square 18SWK)
.PP
.Vb 7
\&   Planimeter \-\-geoconvert\-input <<EOF
\&   18n 500000 4400000
\&   18n 600000 4400000
\&   18n 600000 4500000
\&   18n 500000 4500000
\&   EOF
\&   => 4 400139.532959 10007388597.2
.Ve
.PP
The following code takes the output from gdalinfo and reports the area
covered by the data (assuming the edges of the image are geodesics).
.PP
.Vb 10
\&   #! /bin/sh
\&   grep \-E \*(Aq^((Upper|Lower) (Left|Right)|Center) \*(Aq |
\&   sed \-e \*(Aqs/d /d/g\*(Aq \-e "s/\*(Aq /\*(Aq/g" | tr \-s \*(Aq(),\er\et\*(Aq \*(Aq \*(Aq | awk \*(Aq{
\&       if ($1 $2 == "UpperLeft")
\&           ul = $6 " " $5;
\&       else if ($1 $2 == "LowerLeft")
\&           ll = $6 " " $5;
\&       else if ($1 $2 == "UpperRight")
\&           ur = $6 " " $5;
\&       else if ($1 $2 == "LowerRight")
\&           lr = $6 " " $5;
\&       else if ($1 == "Center") {
\&           printf "%s\en%s\en%s\en%s\en\en", ul, ll, lr, ur;
\&           ul = ll = ur = lr = "";
\&       }
\&   }
\&   \*(Aq | Planimeter | cut \-f3 \-d\*(Aq \*(Aq
.Ve
.SH ACCURACY
.IX Header "ACCURACY"
Using the \fB\-G\fR option (the default), the accuracy was estimated by
computing the error in the area for 10^7 approximately regular
polygons on the WGS84 ellipsoid.  The centers and the orientations of
the polygons were uniformly distributed, the number of vertices was
log\-uniformly distributed in [3, 300], and the center to vertex
distance log\-uniformly distributed in [0.1 m, 9000 km].
.PP
The maximum error in the perimeter was 200 nm, and the maximum error
in the area was
.PP
.Vb 4
\&   0.0013 m^2 for perimeter < 10 km
\&   0.0070 m^2 for perimeter < 100 km
\&   0.070 m^2 for perimeter < 1000 km
\&   0.11 m^2 for all perimeters
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeoConvert\fR\|(1), \fBGeodSolve\fR\|(1).
.PP
An online version of this utility is available at
<https://geographiclib.sourceforge.io/cgi\-bin/Planimeter>.
.PP
The algorithm for the area of geodesic polygon is
given in Section 6 of C. F. F. Karney, \fIAlgorithms for geodesics\fR,
J. Geodesy 87, 43\-55 (2013);
DOI <https://doi.org/10.1007/s00190\-012\-0578\-z>;
addenda: <https://geographiclib.sourceforge.io/geod\-addenda.html>.
.PP
The algorithm for the area of a rhumb polygon is given in Section 3 of
C. F. F. Karney, \fIThe area of rhumb polygons\fR,
Stud. Geophys. Geod. 68(3\-\-4), 99\-\-120 (2024);
DOI <https://doi.org/10.1007/s11200\-024\-0709\-z>.
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBPlanimeter\fR was written by Charles Karney.
.SH HISTORY
.IX Header "HISTORY"
\&\fBPlanimeter\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 1.4.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/Planimeter.cpp
/**
 * \file Planimeter.cpp
 * \brief Command line utility for measuring the area of geodesic polygons
 *
 * Copyright (c) Charles Karney (2010-2023) <karney@alum.mit.edu> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="Planimeter.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/PolygonArea.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/GeoCoords.hpp>
#include <GeographicLib/AuxLatitude.hpp>

#if defined(_MSC_VER)
// Squelch warnings about constant conditional expressions
#  pragma warning (disable: 4127)
#endif

#include "Planimeter.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    typedef Math::real real;
    Utility::set_digits();
    enum { GEODESIC, AUTHALIC, RHUMB, EXACT };
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f();
    bool reverse = false, sign = true, polyline = false, longfirst = false,
      exact = false, geoconvert_compat = false;
    int linetype = GEODESIC;
    int prec = 6;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-r")
        reverse = !reverse;
      else if (arg == "-s")
        sign = !sign;
      else if (arg == "-l")
        polyline = !polyline;
      else if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-G")
        linetype = GEODESIC;
      else if (arg == "-Q")
        linetype = AUTHALIC;
      else if (arg == "-R")
        linetype = RHUMB;
      else if (arg == "-E")
        exact = true;
      else if (arg == "--geoconvert-input")
        geoconvert_compat = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    // At start linetype is one of GEODESIC, AUTHALIC, RHUMB
    const AuxLatitude ellip(a, f);
    if (linetype == AUTHALIC) {
      linetype = exact ? EXACT : GEODESIC;
      using std::sqrt;
      a = sqrt(ellip.AuthalicRadiusSquared(exact));
      f = 0;
    } else if (linetype == GEODESIC && exact)
      linetype = EXACT;
    // Now linetype is one of GEODESIC, EXACT, RHUMB
    const Geodesic geod(a, f);
    const GeodesicExact geode(a, linetype == EXACT ? f : 0);
    const Rhumb rhumb(a, f, exact && linetype == RHUMB);
    PolygonArea poly(geod, polyline);
    PolygonAreaExact polye(geode, polyline);
    PolygonAreaRhumb polyr(rhumb, polyline);
    GeoCoords p;

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    std::string s, eol("\n");
    real perimeter, area;
    unsigned num;
    std::istringstream str;
    std::string slat, slon, junk;
    real lat = 0, lon = 0;
    while (std::getline(*input, s)) {
      if (!cdelim.empty()) {
        std::string::size_type m = s.find(cdelim);
        if (m != std::string::npos) {
          eol = " " + s.substr(m) + "\n";
          s = s.substr(0, m);
        }
      }
      bool endpoly = s.empty();
      if (!endpoly) {
        try {
          using std::isnan;
          if (geoconvert_compat) {
            p.Reset(s, true, longfirst);
            lat = p.Latitude(); lon = p.Longitude();
          } else {
            str.clear(); str.str(s);
            if (!(str >> slat >> slon))
              throw GeographicErr("incomplete input");
            if (str >> junk)
              throw GeographicErr("extra input");
            DMS::DecodeLatLon(slat, slon, lat, lon, longfirst);
          }
          if (isnan(lat) || isnan(lon))
            endpoly = true;
        }
        catch (const GeographicErr&) {
          endpoly = true;
        }
      }
      if (endpoly) {
        num =
          linetype == EXACT ? polye.Compute(reverse, sign, perimeter, area) :
          linetype == RHUMB ? polyr.Compute(reverse, sign, perimeter, area) :
          poly.Compute(reverse, sign, perimeter, area); // geodesic + authalic
        if (num > 0) {
          *output << num << " " << Utility::str(perimeter, prec);
          if (!polyline) {
            *output << " " << Utility::str(area, std::max(0, prec - 5));
          }
          *output << eol;
        }
        linetype == EXACT ? polye.Clear() :
          linetype == RHUMB ? polyr.Clear() : poly.Clear();
        eol = "\n";
      } else {
        linetype == EXACT ? polye.AddPoint(lat, lon) :
          linetype == RHUMB ? polyr.AddPoint(lat, lon) :
          poly.AddPoint
          (linetype == AUTHALIC ?
           ellip.Convert(AuxLatitude::PHI, AuxLatitude::XI, lat, exact) : lat,
           lon);
      }
    }
    num =
      linetype == EXACT ? polye.Compute(reverse, sign, perimeter, area) :
      linetype == RHUMB ? polyr.Compute(reverse, sign, perimeter, area) :
      poly.Compute(reverse, sign, perimeter, area);
    if (num > 0) {
      *output << num << " " << Utility::str(perimeter, prec);
      if (!polyline) {
        *output << " " << Utility::str(area, std::max(0, prec - 5));
      }
      *output << eol;
    }
    linetype == EXACT ? polye.Clear() :
      linetype == RHUMB ? polyr.Clear() : poly.Clear();
    eol = "\n";
    return 0;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
